[Chen Mingyi] iP

mingyi456 opened the pull request:
# Dude, the best enhancement to your existing project

Merging this pull request does the following:

- Changes the boring old name "Duke" to the exciting, **NEW** name "[Dude](https://github.com/mingyi456/ip)" :+1:
- Implements features like <ins>seaching</ins> for a keyword amonst existing tasks :+1:

Simply do the following:

1. Click the "Merge Pull Request" button :+1:
2. Build the new code in your IDE of choice :+1:
3. Enjoy the new enchancements brought to you courtesy of Dude! :+1:

WIP:
- [ ] Make this enhancement even more awesome :+1:
- [x] Convince you to merge this pull request :+1:

Commit:
Add Gradle support

mingyi456 committed:
Level-1 commit

mingyi456 committed:
Complete requirements for Level-2

mingyi456 committed:
Complete Level-3 tasks, and remove unnecessary code in ToDoList.java

mingyi456 committed:
Complete Level-4 tasks

mingyi456 committed:
Complete A-TextUiTesting tasks

mingyi456 committed:
Attempt Level-5 tasks

mingyi456 committed:
Complete Level-6 tasks.

mingyi456 committed:
Add support for saving current task list into data file

mingyi456 committed:
Store dates as java.time.LocalDate objects

mingyi456 committed:
Merge branch 'branch-Level-7'

mingyi456 committed:
Merge branch 'branch-Level-8'

mingyi456 committed:
Add more classes for future refactoring into a more OOP approach

mingyi456 committed:
Organise all classes into a single package duke

mingyi456 committed:
Add JUnit tests for Duke and DateString classes

mingyi456 committed:
Add manifest.mf into .gitignore

mingyi456 committed:
Add a few JavadDoc comments. Will add more in future.

mingyi456 committed:
Implement the find() method for tasks

mingyi456 committed:
Added more JavaDocs

mingyi456 committed:
Ensure existing code adheres to a standard.

mingyi456 committed:
Merge branch 'branch-A-JavaDoc'

mingyi456 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	duke_data.txt
#	src/main/java/duke/Duke.java

mingyi456 committed:
Resolve the merge conflict issue in the previous commit

mingyi456 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/TaskList.java

Jingwencloud commented on src/main/java/duke/DateString.java:
```diff
@@ -0,0 +1,46 @@
+    private LocalDate localDate;
+
+    /**
+     * Constructor which takes in a string and attempts to parse it to create a LocalDate
+     * object. If the parsing is unsuccessful, silently defaults to using a string instead.
+     * @param str
```
Leaving a line between the description and parameters section could improve readability. Not sure if this is intended as I notice you have left a blank line for the other javadocs.

Jingwencloud commented on src/main/java/duke/DateString.java:
```diff
@@ -0,0 +1,46 @@
+     *
+     * @return The string representation of this object.
+     */
+
```
Removing this blank line would comply more with the coding standard.

Jingwencloud commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,118 @@
+     * @param str The line of input entered by the user via stdin
+     * @param taskList The instance of Tasklist used to store the current tasks
+     */
+    static boolean processInput(String str, TaskList taskList) {
```
Maybe you could use a method name that sounds more like a boolean?

tsy24 replied:
Seems like the boolean returned from processInput is not used in the flow of the program. If that is the case, then perhaps the method can return void?

mingyi456 replied:
Well, I struggled quite a while trying to come up with a better implementation for this, but evidently I failed. The `boolean` value is assigned to `isRunning` in `Duke.main()`, and only returns `false` if the command to exit is detected. This is a really "hacky" solution and easily leads to confusion, as can be seen here. I will look into ways to resolve this in future.

Jingwencloud commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,118 @@
+        return sc.nextLine();
+    }
+
+    /**
+     * Contains all the input parsing code currently, will be refactored in the future.
+     *
+     * @param str The line of input entered by the user via stdin
+     * @param taskList The instance of Tasklist used to store the current tasks
+     */
```
You could add a line for @return to make the method easier to understand for readers.

Jingwencloud reviewed (commented):
LGTM! Just a few coding standard violations, but overall good job on your work :)

tsy24 commented on src/main/java/duke/DataFile.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.Scanner;
+import java.util.ArrayList;
+
+public class DataFile {
```
Consider adding JavaDocs for the DataFile class and its public methods so that others can understand the methods more easily.

tsy24 commented on src/main/java/duke/DateString.java:
```diff
@@ -0,0 +1,46 @@
+     * @param str
+     */
+    public DateString(String str) {
+        this.dateString = str;
```
An advanced coding standard that you can consider following is to avoid unnecessary use of `this` with fields. The `this` keyword should only be used when a field is shadowed by a method or constructor parameter. You can refer to the standards [here](https://se-education.org/guides/conventions/java/index.html)

tsy24 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,46 @@
+            // System.out.println(taskList.addToDo(substr));
+        } else if (input.startsWith("deadline")) {
+            String substr = input.replaceFirst("deadline", "").stripLeading();
+            String[] substrArray = substr.split(" /by ", 2);
```
Perhaps substrings is a better name than substrArray? Using plural form will distinguish clearly between single-valued and multi-valued variables.

mingyi456 replied:
Nice catch on this, I did this before I read the guidelines and did tutorial 1, and took my existing code for granted.

tsy24 reviewed (commented):
Code quality for naming looks good to me! It would be better if you are able to write JavaDocs for the other public methods and classes. Keep up the good work!

mingyi456 committed:
Add the use of enum for determining the type of Task object

mingyi456 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

# Conflicts:
#	text-ui-test/runtest.sh

mingyi456 commented:
Thank you for your kind comments, I really expected quite a lot more critisism considering my lack of polish in my code.

mingyi456 committed:
Fix issues with grade build and run

mingyi456 committed:
Fix typo in JavaDoc comment

mingyi456 committed:
Integrate JavaFX into application

A GUI interface is the next intended level for this iP, and it has been
overdue for quite some time

Currently not fully functional, will be improved on soon

mingyi456 committed:
Add detection of empty inputs into GUI interface

mingyi456 committed:
Merge branch 'branch-A-Gradle'

# Conflicts:
#	duke_data.txt

mingyi456 committed:
Merge branch 'branch-Level-10'

mingyi456 committed:
Implement CLI functionality into GUI interface

Previously, the GUI interface only echoed back user input, which is
inconsistent with the CLI interface behaviour.

The CLI functionality is now ported over to the GUI interface

mingyi456 committed:
Use Java assertions to enforce critical assumptions in internal logic

Printing error messages (removed from production code),
or using a boolean return value (used in Task.markDone()) was used,
which is inconvenient (deleting the print statements) and restrictive
(markDone() cannot be void or return some other type)

Java assertions were used to replace these "hacky" workarounds

This improves code flexibility and allows for more convenience, as the
assert statements can be left in the code and either enabled or disabled
at runtime

mingyi456 committed:
Add code quality improvements throughtout the codebase

mingyi456 committed:
Merge pull request #1 from mingyi456/branch-A-Assertions

Use Java assertions to enforce critical assumptions in internal logic

mingyi456 committed:
Merge branch 'master' of https://github.com/mingyi456/ip

* 'master' of https://github.com/mingyi456/ip:
  Use Java assertions to enforce critical assumptions in internal logic

mingyi456 committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/GUI.java

mingyi456 committed:
Merge pull request #2 from mingyi456/branch-A-CodeQuality

Add code quality improvements throughtout the codebase

mingyi456 committed:
Merge branch 'master' of https://github.com/mingyi456/ip

* 'master' of https://github.com/mingyi456/ip:
  Add code quality improvements throughtout the codebase

mingyi456 committed:
Minor tweaks to GUI and add template for more JUnit tests

mingyi456 committed:
Update README.md

mingyi456 committed:
Add placeholder UI.png

mingyi456 committed:
Merge branch 'master' of https://github.com/mingyi456/ip

* 'master' of https://github.com/mingyi456/ip:
  Update README.md

mingyi456 committed:
Rename UI.png to Ui.png to adhere to grading scripts

mingyi456 committed:
Add more JavaDocs comments and tweak resource dependencies for shadowJar

mingyi456 committed:
Add more JavaDocs comments.

mingyi456 committed:
Refactor unnecessary code and add more JavaDocs

mingyi456 committed:
Tweak the command list to make it more intuitive

mingyi456 committed:
Remove more unnecessary code

mingyi456 committed:
Update README.md

mingyi456 committed:
Update README.md

mingyi456 committed:
Merge branch 'master' of https://github.com/mingyi456/ip

* 'master' of https://github.com/mingyi456/ip:
  Update README.md
  Update README.md

mingyi456 committed:
Update README.md

mingyi456 committed:
Update README.md

mingyi456 committed:
Merge branch 'master' of https://github.com/mingyi456/ip

* 'master' of https://github.com/mingyi456/ip:
  Update README.md

mingyi456 committed:
Update README.md

mingyi456 committed:
Update README.md

mingyi456 committed:
Merge branch 'master' of https://github.com/mingyi456/ip

* 'master' of https://github.com/mingyi456/ip:
  Update README.md
  Update README.md

mingyi456 committed:
Update README.md

mingyi456 committed:
Rename duke_data.txt to dude_data.txt

mingyi456 committed:
Merge branch 'master' of https://github.com/mingyi456/ip

* 'master' of https://github.com/mingyi456/ip:
  Update README.md

mingyi456 committed:
Remove dead code in Parser.java

mingyi456 committed:
Update README.md

mingyi456 committed:
Merge branch 'master' of https://github.com/mingyi456/ip

* 'master' of https://github.com/mingyi456/ip:
  Update README.md

mingyi456 committed:
Add more JavaDocs

mingyi456 committed:
Remove more dead code

mingyi456 committed:
Add more JavaDocs

mingyi456 committed:
Test commit

mingyi456 committed:
Revert "Test commit"

This reverts commit 857aa1be7ea49f2e4ca597fd28c509c19dcb1c92.
